main/webservices/api/test_api.php

```diff
@@ -2,116 +2,531 @@
+    /**
+     * Posts an action request to the web server, asserts it returns a valid JSON-encoded response and decodes it
+     * supplied parameters complete or override the generated base parameters username, api_key and action
+     *
+     * @param string    $action         name of the webservice action to call
+     * @param array     $parameters     parameters to send with the request as an associative array
+     * @return mixed                    the decoded response (usually an object with properties data, error, message)
+     */
+    protected function decodedResponse($action, $parameters = [])
+    {
+        $baseParams = [
+            'username' => self::WEBSERVICE_USERNAME,
+            'api_key' => $this->apiKey,
+            'action' => $action,
+        ];
+        $response = $this->client->post(self::URI, ['form_params' => array_merge($baseParams, $parameters)]);
+        $this->assertNotNull($response);
+        $this->assertSame(200, $response->getStatusCode());
+        $decodedResponse = json_decode($response->getBody()->getContents());
+        $this->assertNotNull($decodedResponse);
+        return $decodedResponse;
+    }
 
-    if ($jsonResponse->error) {
-        throw new Exception('Authentication failed because : ' . $jsonResponse->message);
+    /**
+     * Posts an action request and assert the server returns an error message
+     *
+     * @param string    $action         name of the webservice action
+     * @param array     $parameters     parameters to send with the request
+     * @return string                   the error string returned by the webservice
+     */
+    protected function errorMessageString($action, $parameters = [])
+    {
+        $decodedResponse = $this->decodedResponse($action, $parameters);
+        $this->assertIsObject($decodedResponse);
+        $this->assertTrue(property_exists($decodedResponse, 'error'));
+        $error = $decodedResponse->error;
+        $this->assertIsBool(true, $error);
+        $this->assertTrue($error, 'error is not true: '.print_r($decodedResponse, true));
+        $this->assertTrue(property_exists($decodedResponse, 'message'));
+        $message = $decodedResponse->message;
+        $this->assertIsString($message);
+        return $message;
     }
 
-    return $jsonResponse->data->apiKey;
-}
+    /**
+     * Posts an action request and assert the server returns a "data" array
+     *
+     * @param string    $action         name of the webservice action
+     * @param array     $parameters     parameters to send with the request
+     * @return array                    the "data" array returned by the webservice
+     */
+    protected function dataArray($action, $parameters = [])
+    {
+        $decodedResponse = $this->decodedResponse($action, $parameters);
+        $this->assertIsObject($decodedResponse);
+        $this->assertTrue(property_exists($decodedResponse, 'data'), 'response data property is missing: '.print_r($decodedResponse, true));
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Line exceeds maximum limit of 120 characters; contains 140 characters](https://app.codacy.com/app/chamilo/chamilo-lms/pullRequest?prid=4554450)